src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -108,7 +108,7 @@ CheckForDistributedDeadlocks(void)
 	TransactionNode *transactionNode = NULL;
 	int edgeCount = 0;
 	int localGroupId = GetLocalGroupId();
-	List *workerNodeList = ActiveReadableNodeList();
+	List *workerNodeList = ActiveReadableWorkerNodeList();
```

Comment:
Distributed deadlock detection runs on the coordinator and I think it also checks the local locks, so we are not skipping the coordinator. I added an isolation test for this. Am I missing something?

Reply:
The deadlock detection logic only considers transactions that are originated on the node doing the detection: https://github.com/citusdata/citus/blob/94a7e6475ccdb63b88bcce4472ad0eb9fda04059/src/backend/distributed/transaction/distributed_deadlock_detection.c#L145-L150

So, if queries fired from worker nodes deadlock with each other including the coordinator, we cannot detect by the deadlock detector on the coordinator.

Reply:
We should include the coordinator, both in the locks (does it become the first worker now?) and in the deadlock detection.

We do not prevent shards from being placed on the coordinator, and inserts can conflict, and all_modifications_commutative may be enabled.

Reply:
> We should include the coordinator, both in the locks (does it become the first worker now?)

We're still acquiring the lock on the first worker node. Does it really matter where we acquire the lock? Note that the writes on the coordinator also acquires the lock on the first worker node.

Reply:
> We do not prevent shards from being placed on the coordinator,

The coordinator node is marked `shouldhaveshards = false`. Doesn't that prevent non-reference table shards on the coordinator?

Reply:
I'd be happy if we can conclude the discussion because this seems to the blocker for now:

1) I think as @marcocitus we should acquire the locks on the coordinator for `TRUNCATE reference_table` at `AcquireDistributedLockOnRelations `

2) As @marcocitus suggested, we might get into distributed deadlocks happening on the coordinator because of `citus.all_modifications_commutative`  (though I couldn't reproduce, it should be possible)

3) On MX, citus acquires executor locks on the first worker node for reference table modifications (`AcquireExecutorShardLocks()`->`AcquireExecutorShardLockForRowModify()` -> ` AcquireExecutorShardLockForRowModify()` -> `SerializeNonCommutativeWrites()` -> `LockShardListResourcesOnFirstWorker()` -> `SELECT lock_shard_resources()`). **The modifications on the coordinator still acquires the executor lock on the first worker node. I don't think we need to change that, it should be fine that we still acquire the lock on the first worker.**

Reply:
> We're still acquiring the lock on the first worker node. Does it really matter where we acquire the lock? Note that the writes on the coordinator also acquires the lock on the first worker node.

No. I was mostly saying that we should make sure workers include the coordinator if it can be the first worker, but excluding the coordinator from this list is fine.

Reply:
> The coordinator node is marked shouldhaveshards = false. Doesn't that prevent non-reference table shards on the coordinator?

No, "shoudhaveshards=false" means it should not have shards, not that it does not have shards.